### PR TITLE
Add ArgumentValue#default_used? and Arguments#default_used?

### PR DESCRIPTION
**Pull request by chrisbutcher**
As discussed the other day in the GraphQL #ruby Slack channel, I've been looking for a way, via instrumentation http://graphql-ruby.org/queries/instrumentation.html in an `after_query` method, to determine whether an `GraphQL::Argument` defined on a `GraphQL::InputObjectType` _that has a default_ actually used the default during resolution.

Argument values can be of type `GraphQL::Language::Nodes::VariableIdentifier` when the input object value is defined in a variable, or of the type `GraphQL::Language::Nodes::InputObject` when its defined inline. I could traverse through either to determine if a given argument fell back to its default that way.

But it seems like instrumentation code that seeks to know if argument defaults were used would be greatly simplified if we simply added this detail to ArgumentValue.

@rmosolgo I took your suggestion:

> makes sense to me, maybe `arguments_class.new(...)` could take a new parameter of names where the default was applied?

... and came up with a working solution, but would like some input on my approach and naming of things.

## tldr
With this change, arguments specified either inline or in variables that end up falling back to their definition's default are marked as such through the new methods `GraphQL::Query::ArgumentValue#default_used?` and `GraphQL::Query::Arguments#default_used?(key)`.

Also note that the `assert_nil` warning surfaced when running ArgumentsSpec is fixed in https://github.com/rmosolgo/graphql-ruby/pull/1140

Closed https://github.com/rmosolgo/graphql-ruby/pull/1141 in favour of this PR

cc: @eapache @swalkinshaw

### CODE REVIEW

File: lib/graphql/query/arguments.rb [hunk 1]
```diff
@@ -115,7 +133,7 @@ def wrap_value(value, arg_defn_type)
             wrap_value(value, arg_defn_type.of_type)
           when GraphQL::InputObjectType
             if value.is_a?(Hash)
-              arg_defn_type.arguments_class.new(value)
+              arg_defn_type.arguments_class.new(value, Hash.new(false))
```
Comment (chrisbutcher) [inline lib/graphql/query/arguments.rb]: I think this is a sensible value to pass back to the initializer if an argument default happens to be a Hash.

Comment (rmosolgo) [reply lib/graphql/query/arguments.rb]: Yeah, it makes sense to me, too. But it seems like this code is a bug waiting to happen, huh? I mean, since default values are applied elsewhere, I wonder if there's some case where a default value isn't properly applied. :S 

Anyhow, that doesn't have to do with these changes!

**Comment by chrisbutcher** [comment]
Thanks @rmosolgo!

> do you think we could express that idea with an Array or Set? Personally, I find it a bit clearer, since we're not actually storing anything in the hash's values. What do you think about that change?

💯 in fact my first approach was to use a `Set` for this very reason. I'll switch back to that implementation.

File: lib/graphql/query/arguments.rb [hunk 2]
```diff
@@ -85,7 +95,7 @@ def each_value
         end
       end
 
-      NO_ARGS = self.new({}, argument_definitions: [])
+      NO_ARGS = self.new({}, argument_definitions: [], defaults_used: {})
```
Comment (rmosolgo) [inline lib/graphql/query/arguments.rb]: Should be `Set.new`? I guess it's splitting hairs since `#include?` will still return false..

File: spec/graphql/query/arguments_spec.rb [hunk 3]
```diff
@@ -306,7 +328,7 @@
       assert_equal nil, input_object.arguments_class
 
       GraphQL::Query::Arguments.construct_arguments_class(input_object)
-      args = input_object.arguments_class.new({foo: 3, bar: -90})
+      args = input_object.arguments_class.new({foo: 3, bar: -90}, {})
```
Comment (rmosolgo) [inline spec/graphql/query/arguments_spec.rb]: Also `Set.new`?

**Comment by chrisbutcher** [comment]
(Forgot to ping again after new commit) Updated!

**Comment by rmosolgo** [comment]
Thanks for making this happen! 🎉

**Comment by chrisbutcher** [comment]
Thank _you_, Robert!
